qtwidgetsextra/filesystem/QFileSystemTreeView.cpp

```diff
@@ -0,0 +1,242 @@
+#include "QFileSystemTreeView.h"
+
+#include <QFileSystemModel>
+#include <QHeaderView>
+
+class QFileSystemTreeViewPrivate : public QObject
+{
+    Q_OBJECT
+
+public:
+    QFileSystemTreeViewPrivate(QFileSystemTreeView *widgetP)
+        : QObject(widgetP)
+        , widget(widgetP)
+        , model(new QFileSystemModel(this)) {
+        Q_ASSERT(widget);
+    }
+
+    void updateItemSelection(const QStringList &selection, QItemSelectionModel::SelectionFlags flags, int column = 0) const {
+        // QMap give us sorting for free
+        QMap<QModelIndex, QMap<QModelIndex, int> > mapIndexes;
+        QItemSelection itemSelection;
+
+        foreach (const QString &item, selection) {
+            const QModelIndex index = model->index(item, column);
+
+            if (index.isValid()) {
+                mapIndexes[index.parent()][index]++;
+            }
+        }
+
+        foreach (const QModelIndex &parent, mapIndexes.keys()) {
+            const QModelIndexList indexes = mapIndexes[parent].keys();
+            int top = -1;
+            int bottom = -1;
+
+            for (int i = 0; i < indexes.count(); i++) {
+                const QModelIndex &index = indexes[i];
+                const bool hasNext = i +1 < indexes.count();
+
+                if (top ==-1) {
+                    top = index.row();
+                }
+                else {
```

Comment:
five indentation level from the foreach to the last if/else. Can't do something better less deep?